### PR TITLE
add react-recompose types

### PR DESCRIPTION
**Pull request by daleyjem**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

File: types/react-recompose/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,716 @@
...
+// Project: https://github.com/react-recompose/react-recompose
+// Definitions by: Iskander Sierra <https://github.com/iskandersierra>
+//                 Curtis Layne <https://github.com/clayne11>
+//                 Rasmus Eneman <https://github.com/Pajn>
+//                 Lucas Terra <https://github.com/lucasterra>
+//                 Mathieu Masy <https://github.com/TiuSh>
+//                 Chris Brody <https://github.com/brodybits>
+//                 Jeremy Daley <https://github.com/daleyjem>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.8
```
Comment (brodybits) [inline types/react-recompose/index.d.ts:11]: nit: looks like a quite old version of TypeScript

**Comment by brodybits** [comment]
@daleyjem thanks for the proposal! I would love to review, unfortunately cannot promise when I may have time due to other existing commitments. Please feel free to bug me here or in <https://github.com/react-recompose/react-recompose/issues/32>.

**Comment by daleyjem** [comment]
> @daleyjem thanks for the proposal! I would love to review, unfortunately cannot promise when I may have time due to other existing commitments. Please feel free to bug me here or in [react-recompose/react-recompose#32](https://github.com/react-recompose/react-recompose/issues/32).

This for the most part is a copy/paste renaming from `recompose`. As you know, the current recommendation in `react-recompose` is to alias itself, and use `@types/recompose`... so in theory, if those are good enough, then these should be too.

If/when `react-recompose` gets its own type definitions, then these types should be removed... but as of now, I certainly appreciate any help to get this alternative available. Ultimately, I think approval will need to come from a maintainer.

**Review by gabritto — APPROVED** [review]

**Comment by daleyjem** [comment]
Ready to merge

**Comment by daleyjem** [comment]
Thanks so much, @gabritto !!

**Pull request merged**